packages/profiling-node/src/integration.ts

```diff
@@ -1,31 +1,334 @@
+        DEBUG_BUILD &&
+          logger.log('[Profiling] max profile duration elapsed, stopping profiling for:', spanToJSON(span).description);
+
+        const profile = stopSpanProfile(span, profile_id);
+        if (profile) {
+          addToProfileQueue(profile_id, profile);
+        }
+      }, maxProfileDurationMs);
+
+      // Unref timeout so it doesn't keep the process alive.
+      timeout.unref();
+
+      getCurrentScope().setContext('profile', { profile_id });
+      spanToProfileIdMap.set(span, profile_id);
+    }
+  });
+
+  client.on('spanEnd', span => {
+    const profile_id = spanToProfileIdMap.get(span);
+
+    if (profile_id) {
+      if (PROFILE_TIMEOUTS[profile_id]) {
+        global.clearTimeout(PROFILE_TIMEOUTS[profile_id]);
+        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+        delete PROFILE_TIMEOUTS[profile_id];
+      }
+      const profile = stopSpanProfile(span, profile_id);
+
+      if (profile) {
+        addToProfileQueue(profile_id, profile);
+      }
+    }
+  });
+
+  client.on('beforeEnvelope', (envelope): void => {
+    // if not profiles are in queue, there is nothing to add to the envelope.
+    if (!PROFILE_MAP.size) {
+      return;
+    }
+
+    const profiledTransactionEvents = findProfiledTransactionsFromEnvelope(envelope);
+    if (!profiledTransactionEvents.length) {
+      return;
+    }
+
+    const profilesToAddToEnvelope: Profile[] = [];
+
+    for (const profiledTransaction of profiledTransactionEvents) {
+      const profileContext = profiledTransaction.contexts?.['profile'];
+      const profile_id = profileContext?.['profile_id'];
+
+      if (!profile_id) {
+        throw new TypeError('[Profiling] cannot find profile for a transaction without a profile context');
+      }
+
+      // Remove the profile from the transaction context before sending, relay will take care of the rest.
+      if (profileContext) {
+        delete profiledTransaction.contexts?.['profile'];
+      }
+
+      const cpuProfile = takeFromProfileQueue(profile_id);
+      if (!cpuProfile) {
+        DEBUG_BUILD && logger.log(`[Profiling] Could not retrieve profile for transaction: ${profile_id}`);
+        continue;
+      }
+
+      const profile = createProfilingEvent(client, cpuProfile, profiledTransaction);
+      if (!profile) return;
+
+      profilesToAddToEnvelope.push(profile);
+
+      // @ts-expect-error profile does not inherit from Event
+      client.emit('preprocessEvent', profile, {
+        event_id: profiledTransaction.event_id,
+      });
+    }
+
+    addProfilesToEnvelope(envelope, profilesToAddToEnvelope);
+  });
+}
+
+interface ChunkData {
+  id: string | undefined;
+  timer: NodeJS.Timeout | undefined;
+  startTimestampMS: number | undefined;
+  startTraceID: string | undefined;
+}
+class ContinuousProfiler {
+  private _profilerId = uuid4();
+  private _client: NodeClient | undefined = undefined;
+  private _chunkData: ChunkData = {
```

Comment:
m: I think the type for this should actually be `ChunkData | undefined` where `ChunkData` is

```
interface ChunkData {
  id: string;
  timer: NodeJS.Timeout | undefined;
  startTimestampMS: number;
  startTraceID: string;
}
```

Therefore we can do `this._chunkData` checks to early return, and then read from the `this._chunkData.X` properties without further validation.

Reply:
Oh yes, thats even better.